### ISSUE TITLE
prevent tags from being used without classes in scoped style definitions

### ISSUE THREAD

**Issue by indirectlylit** [learningequality/kolibri#6614]
### Observed behavior

In some places in the codebase, we do things like

```html
<template>
  <div>
    <h2>Title</h2>
    <p>paragraph</p>
  </div>
</template>

<style scoped>
  h2 {
    margin-top: 8px;
  }
</style>
```

This has a few issues:

1. According to the [vue docs](https://vue-loader.vuejs.org/guide/scoped-css.html#also-keep-in-mind), using tags inside scoped style blocks "will be many times slower"
2. The DOM structure is coupled to the styling. For example, turning `h2` to `h3` would require changes on multiple lines

### Expected behavior

We should add a [stylelint rule](https://github.com/Moxio/stylelint-selector-tag-no-without-class) to prevent this behavior, and update all existing instances to either use an inline style or a class.

**Inline style:**

```html
<template>
  <div>
    <h2 style="margin-top: 8px;">Title</h2>
    <p>paragraph</p>
  </div>
</template>
```

**Class:**

```html
<template>
  <div>
    <h2 class="title">Title</h2>
    <p>paragraph</p>
  </div>
</template>

<style scoped>
  .title {
    margin-top: 8px;
  }
</style>
```

### User-facing consequences

* slower render performance
* code maintainability

### Context

0.13.x

**Comment by rationality6** [comment]
Can I pick this? I think I could do this.

1. add rule stylelint-selector-tag-no-without-class for prevent commit
2. change all the tags to class or id

**Comment by rationality6** [comment]
Below is my Stylelint set

![Screen Shot 2020-05-05 at 3 56 27 PM](https://user-images.githubusercontent.com/3889468/81041699-6dddd480-8ee9-11ea-9357-8f567eaccaf1.png)

'
npx stylelint '**/*.vue'
'

and these are listing result

![Screen Shot 2020-05-05 at 3 56 38 PM](https://user-images.githubusercontent.com/3889468/81041709-73d3b580-8ee9-11ea-8a8a-aea17a15de4e.png)


Should I change them all to class or id type?

Let me know your guy's opinions and lead me directions.

**Comment by indirectlylit** [comment]
Hi @rationality6 thanks for this! Yes please, let's move forward with this.

In general there will be two options. You can use the following heuristic:

_Does the style have more than one property?_

* If _yes_, use a class with some simple, meaningful name (e.g. `<h1 class="header">`)
* If _no_, convert to an inline style in the template (e.g. `<h1 style="margin: 0">`)

We should never use ID-based styles.

**Comment by bjester** [comment]
I don't think enforcing classes instead of tags will help very much. I still see class styles scoped with a data attribute selector, so I think they're both on the same level regarding data attribute affecting performance.

I found a test suite comparing selectors. Test 4 is tag with attribute selector, and test 20 a single class selector. Unfortunately there's no test for a class with attribute selector. You can see the results in a [table here](https://ecss.io/appendix1.html#h-H2_0), which shows it's pretty negligible, and of course doesn't report on a class + attribute selector.

**Comment by indirectlylit** [comment]
> I still see class styles scoped with a data attribute selector, so I think they're both on the same level regarding data attribute affecting performance.

This is guidance from the official vue docs: https://vue-loader.vuejs.org/guide/scoped-css.html#also-keep-in-mind

Unfortunately the link "Here's a playground where you can test the differences yourself" is broken... If their guidance is no longer accurate though, maybe we should let them know


> I don't think enforcing classes instead of tags will help very much

Regardless of performance I think it still helps with encouraging decoupled CSS, but I agree it's much less important without the performance benefit

**Comment by indirectlylit** [comment]
removing the 0.14 milestone. It was originally targeted because @rationality6 was working on this, but appears that PR is closed

**Comment by bjester** [comment]
> This is guidance from the official vue docs: https://vue-loader.vuejs.org/guide/scoped-css.html#also-keep-in-mind

Yeah I saw that but I think it relies on the condition "will be many times slower when scoped (i.e. when combined with an attribute selector)". From what I see, class selectors become combined with an attribute selector too.

> Unfortunately the link "Here's a playground where you can test the differences yourself" is broken... 

Hmm if you scroll down, the the names of each test are links and launch each test which alerts the time it takes. I observed similar results to the table, although much slower for both.

[Test 4](http://benfrain.com/selector-test/04.html)
[Test 20](http://benfrain.com/selector-test/20.html)
